src/frequenz/sdk/actor/power_distributing/result.py

```diff
@@ -3,122 +3,100 @@
-            request: The user's request to which this message responds.
-            succeed_power: Part of the requested power that was successfully set.
-            succeed_batteries: Subset of the requested batteries for which the request
-                succeed.
-            failed_power: Part of the requested power that failed.
-            failed_batteries: Subset of the requested batteries for which the request
-                failed.
-            excess_power: Part of the requested power that could not be fulfilled,
-                because it was outside available power bounds.
-        """
-        super().__init__(request)
-        self.succeed_power: int = succeed_power
-        self.succeed_batteries: Set[int] = succeed_batteries
-        self.failed_power: int = failed_power
-        self.failed_batteries: Set[int] = failed_batteries
-        self.excess_power: int = excess_power
+@dataclasses.dataclass
+class _BaseResultMixin:
+    """Base mixin class for reporting power distribution results."""
 
+    request: Request
+    """The user's request to which this message responds."""
+
+
+# When moving to Python 3.10+ we should replace this with an union type:
+# Result = Success | PartialFailure | Error | OutOfBound | Ignored
+# For now it can't be done because before 3.10 isinstance(result, Success)
+# doesn't work, so it is hard to figure out what type of result you got in
+# a forward compatible way.
+# When moving we should use the _BaseResultMixin as a base class for all
+# results.
+@dataclasses.dataclass
+class Result(_BaseResultMixin):
+    """Power distribution result."""
+
+
+@dataclasses.dataclass
+class _BaseSuccessMixin:
+    """Result returned when setting the power succeed for all batteries."""
+
+    succeeded_power: int
+    """The part of the requested power that was successfully set."""
+
+    succeeded_batteries: set[int]
+    """The subset of batteries for which power was set successfully."""
+
+    excess_power: int
+    """The part of the requested power that could not be fulfilled.
+
+    This happens when the requested power is outside the available power bounds.
+    """
 
-class Error(Result):
-    """Error occurred and power was not set."""
 
-    def __init__(self, request: Request, msg: str) -> None:
-        """Create class instance.
+# We need to put the _BaseSuccessMixin before Result in the inheritance list to
+# make sure that the Result attributes appear before the _BaseSuccessMixin,
+# otherwise the request attribute will be last in the dataclass constructor
+# because of how MRO works.
 
-        Args:
-            request: The user's request to which this message responds.
-            msg: Error message explaining why error happened.
-        """
-        super().__init__(request)
-        self.msg: str = msg
 
+@dataclasses.dataclass
+class Success(_BaseSuccessMixin, Result):  # Order matters here. See above.
+    """Result returned when setting the power succeed for all batteries."""
```

Comment:
```suggestion
    """Result returned when setting the power has succeeded for all batteries."""
```

Reply:
Here I wanted to put "setting the power succeeded for all batteries."